cpuid/src/brand_string.rs

```diff
@@ -206,50 +206,99 @@ impl BrandString {
     /// Searches the brand string for the CPU frequency data it may contain (e.g. 4.01GHz),
     /// and, if found, returns it as an `u8` slice.
     ///
-    /// Basically, we're implementing a search for this regex: "([0-9]+\.[0-9]+[MG]Hz)".
+    /// Basically, we're implementing a search for this regex: "([0-9]+\.[0-9]+[MGT]Hz)".
     pub fn find_freq(&self) -> Option<&[u8]> {
-        let mut it = self
-            .as_bytes()
-            .iter()
-            .rev()
-            .skip(self.as_bytes().len() - self.len)
-            .enumerate();
-        let mut freq_start = 0_usize;
-        let mut freq_end = 0_usize;
-
-        while freq_start == 0 {
-            match it.next() {
-                Some((i, &b'z')) => freq_end = self.len - i - 1,
-                Some((_, _)) => continue,
-                None => break,
-            }
-            match it.next() {
-                Some((_, &b'H')) => {}
-                Some((_, _)) => continue,
-                None => break,
-            }
-            match it.next() {
-                Some((_, &ch)) => {
-                    if ch != b'M' && ch != b'G' {
-                        continue;
+        // The algorithm for matching the regular expression above is based
+        // on a Moore machine, and 'stage' represents the current state of
+        // the machine
+        enum Stages {
+            /// Initial state, looking for a digit.
+            Initial,
+            /// Found integer part of the frequency
```

Comment:
nit: comments should end in fullstop `.` - applicable to all comments added by this commit.